sqlmesh/core/engine_adapter/base.py

```diff
@@ -67,6 +68,32 @@ def is_insert_overwrite(self) -> bool:
     def is_replace_where(self) -> bool:
         return self == InsertOverwriteStrategy.REPLACE_WHERE
 
+    @property
+    def requires_condition(self) -> bool:
+        return self.is_replace_where or self.is_delete_insert
+
+
+@dataclass
+class SourceQuery:
+    generator: t.Iterator[Query]
```

Comment:
Are we expecting the batches to not fit the memory? If not, then why not materialize all of the queries? Why bother with generators?  Especially provided that we loose the re-entrance property. Basically once generator / single-query is used once it becomes useless and the `SourceQuery` is left in an invalid state.

Reply:
Generally, using iterators / generators outside iteration / context manager is a dangerous idea to say the least. Let alone as part of a shared stateful object.

Reply:
generators won't be able to fit in memory because they could be large sql queries, agreed that it should be in a generator obj or iterator as opposed to an instance variable like this

Reply:
SourceQuery should be a base class, and other adapters should inherit it. right now it's strange because you're instantiating it and passing in a "generator" into each one.

Reply:
> agreed that it should be in a generator obj or iterator as opposed to an instance variable like this

It is a generator object? I basically had a `t.Tuple[t.Iterator, <a bunch of other stuff>]` before so instead of doing that I created a dataclass (or Pydantic model doesn't matter) for it. 

> SourceQuery should be a base class, and other adapters should inherit it.

That could be done but I'm not currently seeing why this would be much better. Each class today implements this the same way just with different generators based on how they handle yielding the expressions after converting a dataframe.

Reply:
> generators won't be able to fit in memory because they could be large sql queries, agreed that it should be in a generator obj or iterator as opposed to an instance variable like this

I don't get this. We can fir the dataframe in memory but not SQL queries derived from it?

Reply:
1 millon row data frame fits in memory ez, but you convert that into union all statements, you're in trouble, but maybe that's not what ryan was thinking?

Reply:
With SQL of that size the issue wouldn't be memory of the machine but rather limits on the engine in terms of how big of a SQL statement it will parse. Currently the batching we do with Pandas when generating SQL queries is because of engine limits and not because of local memory limits.